### PR TITLE
[JENKINS-31518] BuildDiscarderProperty settings compatibility

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-31518](https://issues.jenkins-ci.org/browse/JENKINS-31518)

Follow-up to #1889. Maintaining settings compatibility can be tricky! Perhaps we need a general system for preprocessing XML about to be passed to XStream that does not require Java-level linkage or `readResolve` or `ItemListener.onLoaded`, so that new code could be relatively clean and efficient.

@reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/model/Job.java [hunk 1]
```diff
@@ -572,6 +576,13 @@ public void removeProperty(JobProperty<? super JobT> jobProp) throws IOException
      * this job.
      */
     public <T extends JobProperty> T getProperty(Class<T> clazz) {
+        if (clazz == BuildDiscarderProperty.class && logRotator != null) {
+            return clazz.cast(new BuildDiscarderProperty(logRotator));
+        }
+        return _getProperty(clazz);
+    }
+
+    private <T extends JobProperty> T _getProperty(Class<T> clazz) {
```
Comment (KostyaSha) [inline core/src/main/java/hudson/model/Job.java]: Seems it doesn't fit into java naming convention: `Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.`  Maybe call `getProperty2()`?

**Comment by KostyaSha** [comment]
> Maintaining settings compatibility can be tricky! Perhaps we need a general system for preprocessing XML about to be passed to XStream that does not require Java-level linkage or readResolve or ItemListener.onLoaded, so that new code could be relatively clean and efficient.

Do you mean path XmlObject into some `readResolveXml()`?

File: core/src/main/java/hudson/model/Job.java (hunk 1 above)
Comment (jglick) [inline core/src/main/java/hudson/model/Job.java]: It is common to use an underscore prefix to indicate a private variant of an API method. `getPropertyInternal` or `rawGetProperty` etc. would be fine too.

**Comment by jglick** [comment]
> Do you mean provide some XmlObject into some `readResolveXml()`

Something along those lines.

**Comment by jglick** [comment]
Would have liked wider review, but I failed to grab @kohsuke’s attention, and this is critical enough that it should go into 1.639 before he cuts it, so will merge.

File: core/src/main/java/hudson/model/Job.java (hunk 1 above)
Comment (KostyaSha) [inline core/src/main/java/hudson/model/Job.java]: In C - yes.

**Comment by KostyaSha** [comment]
Failed test unrelated?

**Comment by jglick** [comment]
Yes I think so.

**Pull request merged**